symengine/sets.cpp

```diff
@@ -464,4 +554,250 @@ RCP<const Boolean> Union::contains(const RCP<const Basic> &o) const
+    if (c1 != 0)
+        return c1;
+    else
+        return unified_compare(container_, other.container_);
+}
+
+RCP<const Boolean> Complement::contains(const RCP<const Basic> &a) const
+{
+    return logical_and(
+        {universe_->contains(a), logical_not(container_->contains(a))});
+}
+
+RCP<const Set> Complement::set_union(const RCP<const Set> &o) const
+{
+    // A' U C = (A n C')'
+    RCP<const Set> ocomplement = o->set_complement(universe_);
+    RCP<const Set> intersect
+        = SymEngine::set_intersection({container_, ocomplement});
+    return intersect->set_complement(universe_);
+}
+
+RCP<const Set> Complement::set_intersection(const RCP<const Set> &o) const
+{
+    return SymEngine::set_intersection({rcp_from_this_cast<const Set>(), o});
+}
+
+RCP<const Set> Complement::set_complement(const RCP<const Set> &o) const
+{
+    auto newuniv = SymEngine::set_union({o, universe_});
+    return container_->set_complement(newuniv);
+}
+
+RCP<const Set> set_union(const set_set &in, bool solve)
+{
+    set_set input;
+    if (solve == false && in.size() > 1)
+        return make_rcp<const Union>(in);
+    set_basic combined_FiniteSet;
+    for (auto it = in.begin(); it != in.end(); ++it) {
+        if (is_a<FiniteSet>(**it)) {
+            const FiniteSet &other = down_cast<const FiniteSet &>(**it);
+            combined_FiniteSet.insert(other.get_container().begin(),
+                                      other.get_container().end());
+        } else if (is_a<UniversalSet>(**it)) {
+            return universalset();
+        } else if (not is_a<EmptySet>(**it)) {
+            input.insert(*it);
+        }
+    }
+    if (input.empty()) {
+        return finiteset(combined_FiniteSet);
+    } else if (input.size() == 1 && combined_FiniteSet.empty()) {
+        return *input.begin();
+    }
+    // Now we rely on respective containers' own rules
+    // TODO: Improve it to O(log n)
+    RCP<const Set> combined_Rest = finiteset(combined_FiniteSet);
+    for (auto it = input.begin(); it != input.end(); ++it) {
+        combined_Rest = combined_Rest->set_union(*it);
+    }
+    return combined_Rest;
+}
+
+RCP<const Set> set_intersection(const set_set &in, bool solve)
+{
+    if (solve == false)
+        throw std::runtime_error(
+            "Not implemented"); // No intersection class yet
+
+    // https://en.wikipedia.org/wiki/Intersection_(set_theory)#Nullary_intersection
+    if (in.empty())
+        return universalset();
+
+    // Global rules
+    // If found any emptyset then return emptyset
+    set_set incopy;
+    for (const auto &input : in) {
+        if (is_a<EmptySet>(*input))
+            return emptyset();
+        else if (not is_a<UniversalSet>(*input))
+            incopy.insert(input);
+    }
+
+    if (incopy.empty())
+        return universalset();
+    if (incopy.size() == 1)
+        return *incopy.begin();
+
+    // Handle finite sets
+    set_set fsets, othersets;
```

Comment:
Good catch @isuruf . Thanks
Complexity wise vector of sets is a better option as sets take `O(logN)` for insertion.